tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/ScalaJSOptimizer.scala

```diff
@@ -81,29 +84,50 @@ class ScalaJSOptimizer(val semantics: Semantics, val outputMode: OutputMode,
   def optimizeIR(irFiles: Traversable[VirtualScalaJSIRFile],
       cfg: Config, logger: Logger): Unit = {
 
-    val builder = {
-      import cfg._
-      if (wantSourceMap)
-        new JSFileBuilderWithSourceMap(output.name,
-            output.contentWriter,
-            output.sourceMapWriter,
-            relativizeSourceMapBase)
-      else
-        new JSFileBuilder(output.name, output.contentWriter)
-    }
+    optimizeIR { emitJS =>
+      val builder = {
+        import cfg._
+        if (wantSourceMap)
+          new JSFileBuilderWithSourceMap(output.name,
+              output.contentWriter,
+              output.sourceMapWriter,
+              relativizeSourceMapBase)
+        else
+          new JSFileBuilder(output.name, output.contentWriter)
+      }
 
-    emitter.emitCustomHeader(cfg.customOutputWrapper._1, builder)
-    emitter.emitPrelude(builder, logger)
-    optimizeIR(irFiles, cfg, builder, logger)
-    emitter.emitPostlude(builder, logger)
-    emitter.emitCustomFooter(cfg.customOutputWrapper._2, builder)
+      emitter.emitCustomHeader(cfg.customOutputWrapper._1, builder)
+      emitter.emitPrelude(builder, logger)
+      emitJS(builder)
+      emitter.emitPostlude(builder, logger)
+      emitter.emitCustomFooter(cfg.customOutputWrapper._2, builder)
 
-    builder.complete()
-    builder.closeWriters()
+      builder.complete()
+      builder.closeWriters()
+    } { () =>
+      import cfg._
+      output.exists && (!wantSourceMap || output.sourceMapExists)
+    } (irFiles, cfg, logger)
   }
 
   def optimizeIR(irFiles: Traversable[VirtualScalaJSIRFile],
       cfg: OptimizerConfig, builder: JSTreeBuilder, logger: Logger): Unit = {
+    optimizeIR(emitJS => emitJS(builder))(() => false)(irFiles, cfg, logger)
+  }
+
+  /** Run the linker pipeline.
+   *
+   *  This method first checks whether the files have changed from the last run.
+   *  If there is no change, [[hasOldResult]] is called. If it returns true,
+   *  no further action is taken and [[None]] is returned.
+   *  Otherwise, the whole pipeline is run. At the end, [[buildJS]] is called
+   *  with a function that emits the trees to the specified builder. The result
+   *  is wrapped into a [[Some]] and returned.
+   */
+  def optimizeIR[T](buildJS: (JSTreeBuilder => Unit) => T)(
+      hasOldResult: () => Boolean)(
+      irFiles: Traversable[VirtualScalaJSIRFile],
+      cfg: OptimizerConfig, logger: Logger): Option[T] = {
```

Comment:
Consider reordering the parameter lists as follows:
```scala
(irFiles, cfg, logger)(hasOldResult)(buildJS)
```
The rationale is 1. put lambdas at the end, as is more common in Scala APIs and 2. put `hasOldResult` first because it's called first.